### PR TITLE
Thread to scheduler

### PR DESCRIPTION
**Pull request by SamBarker**
Instead of dedicating a single thread to the storage checker and managing it directly use a, potentially shared, Scheduled Executor service to manage the background tasks.
 
Separating this out from the work for https://github.com/strimzi/kafka-quotas-plugin/issues/28 to try and keep the reviews more manageable. 

Having a scheduled executor proved to be quite valuable in the internal version of the quota plugin we developed, and we want to replace with the KIP-827 based version, as it made handling multiple background jobs much simpler. It also improves separation of concerns so that the storage checker is focused on evaluating storage usage rather than thread management.

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java [hunk 1]
```diff
@@ -151,12 +158,17 @@ public void configure(Map<String, ?> configs) {
...
-        log.info("Configured quota callback with {}. Storage quota (soft, hard): ({}, {}). Storage check interval: {}ms", quotaMap, storageQuotaSoft, storageQuotaHard, storageCheckIntervalMillis);
+        if (storageCheckerFuture != null) {
+            storageCheckerFuture.cancel(true);
+        }
+        if (storageCheckIntervalMillis > 0L) {
+            List<Path> logDirs = config.getLogDirs().stream().map(Paths::get).collect(Collectors.toList());
+            storageChecker.configure(
+                    logDirs,
+                    this::updateUsedStorage);
+            storageCheckerFuture = backgroundScheduler.scheduleAtFixedRate(storageChecker, storageCheckIntervalMillis, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
```
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:169]: I'm on the fence about using `atFixedRate` or `withFixedDelay` here. 

I think `fixedRate` matches the configuration intent but `withFixedDelay` would be a closer match to the previous implementation. Given we expect the task to execute very quickly the distinction is probably moot.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 2]
```diff
@@ -4,72 +4,47 @@
...
     private volatile Consumer<Long> consumer;
 
-    void configure(long storageCheckIntervalMillis, List<Path> logDirs, Consumer<Long> consumer) {
-        this.storageCheckIntervalMillis = storageCheckIntervalMillis;
+    void configure(List<Path> logDirs, Consumer<Long> consumer) {
         this.logDirs = logDirs;
         this.consumer = consumer;
     }
 
-    void startIfNecessary() {
```
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:39]: I don't think there is a strong need to maintain the API here but can easily restore the method signatures (and mark deprecated) if desired.

File: src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java [hunk 3]
```diff
@@ -30,10 +24,27 @@
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentCaptor;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (SamBarker) [inline src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java:27]: I don't think there is consistency in the project about where static imports go, but this is at least consistent with `StaticQutoaConfig` and `StaticQutoaCallback`.

File: src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java [hunk 4]
```diff
@@ -30,10 +24,27 @@
...
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
 class StaticQuotaCallbackTest {
 
+    public static final Map<String, Integer> MINIMUM_VALID_CONFIG = Map.of(StaticQuotaConfig.STORAGE_CHECK_INTERVAL_PROP, 10);
```
Comment (SamBarker) [inline src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java:43]: Need to revisit why this change was required.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java (hunk 2 above)
Comment (scholzj) [reply src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:39]: You mean as in _backwards compatibility_? I think this is an implementation detail. The API would be the configuration of the plugin. So I think it is fine to remove this.

**Review by scholzj — COMMENTED** [review]
Out of curiosity ... have you considered `TImer` and `TimerTask` which would abstract the thread completely?

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java (hunk 2 above)
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:39]: Yes, backwards compatibility was my concern.

**Comment by SamBarker** [comment]
Te be honest no, I didn't consider `Timer` & `TimerTask` I went directly to the `ScheduledExecutorService` (I didn't deal with threads in the Java 1.4 days). A quick google suggests the single threaded executor as used here is equivalent and preferred. I will have a deeper look tomorrow to see if I've missed something.

**Comment by scholzj** [comment]
> Te be honest no, I didn't consider Timer & TimerTask I went directly to the ScheduledExecutorService (I didn't deal with threads in the Java 1.4 days). A quick google suggests the single threaded executor as used here is equivalent and preferred. I will have a deeper look tomorrow to see if I've missed something.

I was just curios ... no need to investigate it just because of me.

**Comment by SamBarker** [comment]
Re: `Timer` the javadoc in for JDK 11 says:

> Java 5.0 introduced the java.util.concurrent package and one of the concurrency utilities therein is the [ScheduledThreadPoolExecutor](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledThreadPoolExecutor.html) which is a thread pool for repeatedly executing tasks at a given rate or delay. It is effectively a more versatile replacement for the Timer/TimerTask combination, as it allows multiple service threads, accepts various time units, and doesn't require subclassing TimerTask (just implement Runnable). Configuring ScheduledThreadPoolExecutor with one thread makes it equivalent to Timer.

Which I take to mean this is the recommended way 😁  Using the scheduler would potentially allow a configurable number of threads should that become an issue.

**Review by scholzj — APPROVED** [review]

File: src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java (hunk 4 above)
Comment (SamBarker) [reply src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java]: Naming left something to be desired. An empty map is still "Valid" but doesn't enable the storage checker so some tests failed without it.

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java (hunk 1 above)
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:169]: I'd be inclined to fixed delay myself. That guarantees we can't end up checking continuously. But you're right that it's difficult to see that it matters in normal circumstances.

**Review by tombentley — APPROVED** [review]
A few trivial points, but this LGTM. Thanks!

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java [hunk 5]
```diff
@@ -151,12 +158,17 @@ public void configure(Map<String, ?> configs) {
...
         long storageCheckIntervalMillis = TimeUnit.SECONDS.toMillis(config.getStorageCheckInterval());
-        List<Path> logDirs = config.getLogDirs().stream().map(Paths::get).collect(Collectors.toList());
-        storageChecker.configure(storageCheckIntervalMillis,
-                logDirs,
-                this::updateUsedStorage);
-
-        log.info("Configured quota callback with {}. Storage quota (soft, hard): ({}, {}). Storage check interval: {}ms", quotaMap, storageQuotaSoft, storageQuotaHard, storageCheckIntervalMillis);
+        if (storageCheckerFuture != null) {
+            storageCheckerFuture.cancel(true);
+        }
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:163]: I was just being defensive really as didn't see a reason why the plug-in couldn't implement `Reconfigurable`. As you say we can get rid of the field. I'll remove it.

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java (hunk 1 above)
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java]: I've switched to `withFixedDelay`.

**Comment by SamBarker** [comment]
@tombentley / @scholzj  nits are addressed. Are you happy to merge?

**Comment by scholzj** [comment]
Thanks for the PR.
